pkg/image/resolve.go

```diff
@@ -132,79 +154,86 @@ func walkthroughMap(inputMap map[interface{}]interface{}, walkFunc func(map[inte
 	}
 }
 
-func GetImages(systemChartPath, chartPath string, k3sUpgradeImages, imagesFromArgs []string, rkeSystemImages map[string]rketypes.RKESystemImages, osType OSType) ([]string, error) {
-	var images []string
-
+func GetImages(systemChartPath, chartPath string, k3sUpgradeImages, imagesFromArgs []string, rkeSystemImages map[string]rketypes.RKESystemImages, osType OSType) ([]string, []string, error) {
 	// fetch images from system charts
+	imagesSet := make(map[string]map[string]bool)
 	if systemChartPath != "" {
-		imagesInSystemCharts, err := fetchImagesFromCharts(systemChartPath, osType)
-		if err != nil {
-			return []string{}, errors.Wrap(err, "failed to fetch images from system charts")
+		if err := fetchImagesFromCharts(systemChartPath, osType, imagesSet); err != nil {
+			return nil, nil, errors.Wrap(err, "failed to fetch images from system charts")
 		}
-		images = append(images, imagesInSystemCharts...)
 	}
 
 	// fetch images from charts
 	if chartPath != "" {
-		imagesInCharts, err := fetchImagesFromCharts(chartPath, osType)
-		if err != nil {
-			return []string{}, errors.Wrap(err, "failed to fetch images from charts")
+		if err := fetchImagesFromCharts(chartPath, osType, imagesSet); err != nil {
+			return nil, nil, errors.Wrap(err, "failed to fetch images from charts")
 		}
-		images = append(images, imagesInCharts...)
 	}
 
 	// fetch images from system images
 	if len(rkeSystemImages) > 0 {
-		imagesInSystem, err := fetchImagesFromSystem(rkeSystemImages, osType)
-		if err != nil {
-			return []string{}, errors.Wrap(err, "failed to fetch images from system images")
+		if err := fetchImagesFromSystem(rkeSystemImages, osType, imagesSet); err != nil {
+			return nil, nil, errors.Wrap(err, "failed to fetch images from system images")
 		}
-		images = append(images, imagesInSystem...)
 	}
 
-	// append images from requirement
-	if requirementImages := getRequirementImages(osType); len(requirementImages) > 0 {
-		images = append(images, requirementImages...)
-	}
```

Comment:
why arent we checking length of requirementImages anymore ? Is it because they get checked in `addSourceToImage` ?

Reply:
It no longer returns a slice, it just sets the images in the map if applicable.